File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/procedures_defreturn.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import io.gopiper.piper.engine.model.Func
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeProceduresDefreturn(statement: JSONObject, onError: (e: Exception) -> Unit) {
    try {
        Log.d("hello", "execute_procedures_defreturn")
        val pointer = statement.getJSONObject("pointer")
        val name = pointer["name"].toString()
        val args = pointer.getJSONArray("args")
        val functionBlock = pointer["functionBlock"].toString()
        val ret = pointer["ret"].toString()

        for (i in 0 until args.length()) {
            VARIABLE_STACK[args[i].toString()] = null
        }

        FUNCTION_STACK.add(Func(
            name = name,
            functionBlock = functionBlock,
            ret = ret
        ))
    } catch (e: Exception) {
        onError(e)
    }
}

File: piper/extensions/chrome/src/engine/statements/procedures_defreturn.js
import { stringify } from "../util"

export default async function (statement, onError) {
    try {
        let pointer = statement["pointer"]
        let name = stringify(pointer["name"])
        let args = pointer["args"]
        let functionBlock = stringify(pointer["functionBlock"])
        let ret = stringify(pointer["ret"])

        for (let arg of args) {
            this.VARIABLE_STACK[arg] = null
        }

        this.FUNCTION_STACK.push({
            name: name,
            functionBlock: functionBlock,
            ret: ret
        })
    } catch (e) {
        onError(e)
    }
}